src/ray/gcs/gcs_server/gcs_resource_manager.cc

```diff
@@ -134,14 +140,39 @@ void GcsResourceManager::Initialize(const GcsInitData &gcs_init_data) {
   }
 }
 
-const absl::flat_hash_map<NodeID, ResourceSet> &GcsResourceManager::GetClusterResources()
-    const {
+const absl::flat_hash_map<NodeID, SchedulingResources>
+    &GcsResourceManager::GetClusterResources() const {
   return cluster_scheduling_resources_;
 }
 
-void GcsResourceManager::UpdateResources(const NodeID &node_id,
-                                         const ResourceSet &resources) {
-  cluster_scheduling_resources_[node_id] = resources;
+void GcsResourceManager::SetAvailableResources(const NodeID &node_id,
+                                               const ResourceSet &resources) {
+  cluster_scheduling_resources_[node_id].SetAvailableResources(ResourceSet(resources));
+}
+
+void GcsResourceManager::UpdateResourceCapacity(
+    const NodeID &node_id,
+    const std::unordered_map<std::string, double> &changed_resources) {
+  auto iter = cluster_scheduling_resources_.find(node_id);
+  if (iter != cluster_scheduling_resources_.end()) {
+    SchedulingResources &scheduling_resources = iter->second;
+    for (const auto &entry : changed_resources) {
+      scheduling_resources.UpdateResourceCapacity(entry.first, entry.second);
+    }
+  } else {
+    cluster_scheduling_resources_.emplace(
```

Comment:
Why don't we just call `SetAvailableResources`?

Reply:
`SetAvailableResources` will not set the `resources_total_`  of `SchedulingResources`.